File: Public/Gui/Linn/Gui/Resources/Package.cs
using System.Collections;
using System.Collections.Generic;
using Linn.Gui;
using Linn.Gui.Scenegraph;
using Linn.Gui.Resources;
using System.Xml;
using System.Xml.XPath;
using System.IO;

namespace Linn {
namespace Gui {
namespace Resources {

public sealed class InvalidVersion : System.Exception {
    public InvalidVersion(uint aExpected, uint aActual) : base("Invalid package version: Expected " + aExpected + " got " + aActual) { }
    public InvalidVersion(uint aExpected, uint aActual, System.Exception aExc) : base("Invalid package version: Expected " + aExpected + " got " + aActual, aExc) {}
}

internal sealed class DuplicatePlugin : System.Exception {}
public sealed class InvalidFilename : System.Exception {}
public sealed class RootAlreadyDefined : System.Exception {}
public sealed class NoRootDefined : System.Exception {}
public sealed class NamespaceNotFound : System.Exception {}
    
public sealed class Package : ISerialiseObject
{
    public Package() {
    }
    
    public Package(string aFilename) {
        Trace.WriteLine(Trace.kGui, "Package.Package: Loading package: " + aFilename);
        bool searching = true;
        int pathIndex = 0;
        while(searching) {
            try {
                string filename = Path.Combine(PackageManager.Instance.PathList[pathIndex], aFilename);
                Trace.WriteLine(Trace.kGui, "Package.Package: Searching " + Path.GetDirectoryName(filename) + "...");
                Load(filename);
                iFilename = filename;
                searching = false;
            } catch(System.IO.IOException e) {
                ++pathIndex;
                if(pathIndex == PackageManager.Instance.PathList.Count) {
                    throw e;
                }
            }
        }
    }
    
    public Package(string aDirectory, string aFilename) {
        Trace.WriteLine(Trace.kGui, "Package.Package: Loading package: " + aDirectory + aFilename);
        string filename = Path.Combine(aDirectory, aFilename);
        Load(filename);
        iFilename = filename;
    }
    
    private void Load(string aPathAndFilename) {
        XmlDocument xmlDoc = new XmlDocument();
        xmlDoc.Load(aPathAndFilename);
        XmlNodeList list = xmlDoc.DocumentElement.SelectNodes("/Package");
        if(list != null) {
            Assert.Check(list.Count == 1);
            Load(list[0]);
        }
    }
    
    public void Load(XmlNode aXmlNode) {
        foreach(XmlAttribute attrib in aXmlNode.Attributes) {
            if(attrib.Name == "type") {
                iType = attrib.Value;
            }
        }
        
        XmlNodeList list;
        list = aXmlNode.SelectNodes("Version");
        if(list != null) {
            Assert.Check(list.Count == 1);
            if(kVersion != uint.Parse(list[0].FirstChild.Value)) {
                throw new InvalidVersion(kVersion, uint.Parse(list[0].FirstChild.Value));
            }
        }
        
        list = aXmlNode.SelectNodes("Namespace");
        if(list != null) {
            Assert.Check(list.Count == 1);
            iNamespace = list[0].FirstChild.Value;
        }
        
        list = aXmlNode.SelectNodes("Package");
        if(list != null) {
            foreach(XmlNode node in list) {
                iPackageList.Add(node.FirstChild.Value);
            }
        }
        
        list = aXmlNode.SelectNodes("Plugin");
        if(list != null) {
            foreach(XmlNode node in list) {
                Plugin plugin = PluginFactory.Instance.Create(node);
                plugin.Namespace = iNamespace;
                iPluginList.Add(plugin);
            }
        }
    }
    
    public void Link() {
        Trace.WriteLine(Trace.kGui, "Package: " + Filename + ": linking...");
        for(int i = 0; i < iPluginList.Count; ++i) {
            Plugin p1 = PluginItem(i);
            for(int j = i + 1; j < iPluginList.Count; ++j) {
                Plugin p2 = PluginItem(j);
                if(p1.Fullname == p2.Fullname) {
                    throw new DuplicatePlugin();
                }
            }
        }
        
        foreach(Plugin plugin in iPluginList) {
            plugin.Link();
            NodeHit node = plugin as NodeHit;
            if(node != null) {
                if(node.IsRoot == true) {
                    Trace.WriteLine(Trace.kGui, "Setting " + node.Fullname + " as Root for " + iFilename);
                    Root = node;
                }
            }
        }
        if(Root == null) {
            throw new NoRootDefined();
        }
        Trace.WriteLine(Trace.kGui, "Package: " + Filename + ": linked. root = " + Root.Fullname);
    }
    
    public void Save(XmlTextWriter aWriter) {
        aWriter.WriteStartDocument();
        
        aWriter.WriteStartElement("Package");
        aWriter.WriteAttributeString("type", "Layout");
        aWriter.WriteStartElement("Version");
        aWriter.WriteString(kVersion.ToString());
        aWriter.WriteEndElement();
        aWriter.WriteStartElement("Namespace");
        aWriter.WriteString(iNamespace);
        aWriter.WriteEndElement();
        foreach(string package in iPackageList) {
            aWriter.WriteStartElement("Package");
            //aWriter.WriteString(package.Filename);
            aWriter.WriteString(package);
            aWriter.WriteEndElement();
        }
        foreach(Plugin plugin in iPluginList) {
            aWriter.WriteStartElement("Plugin");
            string type = plugin.GetType().ToString();
            aWriter.WriteAttributeString("class", type.Substring(type.LastIndexOf('.') + 1));
            plugin.Save(aWriter);
            aWriter.WriteEndElement();
        }
        aWriter.WriteEndElement();
        
        aWriter.WriteEndDocument();
    }
    
    public void Save() {
        if(iFilename == "") {
            throw new InvalidFilename();
        }
        Save(iFilename);
    }
    
    public void Save(string aFilename) {
        UpdatePackageDependecies();
        
        Trace.WriteLine(Trace.kGui, "Save with filename=" + aFilename);
        XmlTextWriter xmlWriter = new XmlTextWriter(aFilename, null);
        xmlWriter.Formatting = Formatting.Indented;
        xmlWriter.Indentation = 2;
        Save(xmlWriter);
        xmlWriter.Close();
        iFilename = aFilename;
        /*if(PackageManager.Instance.RootDirectory != "") {
            iFilename = iFilename.Replace(PackageManager.Instance.RootDirectory, "");
            Trace.WriteLine(Trace.kGui, "package filename=" + iFilename + " (" + PackageManager.Instance.RootDirectory + ")");
        }*/
    }
    
    public string Filename {
        get {
            return iFilename;
        }
        set {
            iFilename = value;
        }
    }
    
    public string Namespace {
        get {
            return iNamespace;
        }
        set {
            iNamespace = value;
            foreach(Plugin plugin in iPluginList) {
                plugin.Namespace = iNamespace;
            }
        }
    }
    
    public string Type {
        get {
            return iType;
        }
        set {
            iType = value;
        }
    }
    
    public NodeHit Root {
        get {
            return iRoot;
        }
        set {
            if(iRoot != null) {
                throw new RootAlreadyDefined();
            }
            iRoot = value;
        }
    }
    
    public List<Plugin> PluginList {
        get {
            return iPluginList;
        }
    }
    
    public Plugin PluginItem(int aIndex) {
        return (Plugin)iPluginList[aIndex];
    }
    
    public List<string> PackageList {
        get {
            return iPackageList;
        }
    }
    
    public void FixUpCrossPackageNodes() {
        foreach(Plugin plugin in iPluginList) {
            Node node = plugin as Node;
            if(node != null) {
                if(node.Parent != null) {
                    if(node.Parent.Namespace != Namespace) {
                        node.Active = node.Parent.Active;
                    }
                }
            }
        }
    }
    
    public Plugin PluginByName(string aName) {
        int i = aName.IndexOf('.');
        if(i < 0) {
            throw new NamespaceNotFound();
        }
        string ns = aName.Substring(0, i);
        string n = aName.Substring(i+1);
        if(ns == iNamespace) {
            foreach(Plugin plugin in iPluginList) {
                if(plugin.Name == n) {
                    return plugin;
                }
            }
        }
        //Trace.WriteLine(Trace.kGui, aName + " not found in " + iFilename);
        return null;
    }
    
    public void AddNode(Node aNode) {
        Trace.WriteLine(Trace.kGui, "Package: adding node " + aNode.Fullname + "...");
        AddPlugin(aNode);
        for(int i = 0; i < aNode.Children.Count; ++i) {
            AddNode(aNode.Child(i));
        }
        Trace.WriteLine(Trace.kGui, "Package: node added.");
    }
    
    public void DeleteNode(Node aNode) {
        Trace.WriteLine(Trace.kGui, "Package: deleting node " + aNode.Fullname + "...");
        if(aNode.Parent != null) {
            aNode.Parent.RemoveChild(aNode);
        }
        DeletePlugin(aNode);
        while(aNode.Children.Count > 0) {
            DeleteNode(aNode.Child(0));
        }
        Trace.WriteLine(Trace.kGui, "Package: node deleted.");
    }
    
    public void AddPlugin(Plugin aPlugin) {
        AddPlugin(aPlugin, aPlugin);
    }
    
    private void AddPlugin(Plugin aPlugin, Plugin aStart) {
        aPlugin.Namespace = iNamespace;
        aPlugin.Name = aPlugin.Name;
        iPluginList.Add(aPlugin);
        Trace.WriteLine(Trace.kGui, "Package: added " + aPlugin.Fullname);
        Plugin plugin = aPlugin.NextPlugin;
        if(plugin != null && aStart != plugin) {
            AddPlugin(plugin, aStart);
        }
    }
    
    public void DeletePlugin(Plugin aPlugin) {
        DeletePlugin(aPlugin, aPlugin);
    }
    
    private void DeletePlugin(Plugin aPlugin, Plugin aStart) {
        iPluginList.Remove(aPlugin);
        Trace.WriteLine(Trace.kGui, "Package: deleted " + aPlugin.Fullname);
        Plugin plugin = aPlugin.NextPlugin;
        if(plugin != null && aStart != plugin) {
            DeletePlugin(plugin, aStart);
        }
    }
    
    private void UpdatePackageDependecies() {
        iPackageList.Clear();
        
        List<string> depsList = new List<string>();
        foreach(Plugin p in iPluginList) {
            List<string> deps = p.PackageDependencies();
            foreach(string s in deps) {
                if(!depsList.Contains(s)) {
                    Trace.WriteLine(Trace.kGui, "Adding " + s);
                    depsList.Add(s);
                }
            }
            Plugin n = p;
            while(n != null && n != p) {
                if(n.Namespace != Namespace) {
                    deps = p.PackageDependencies();
                    foreach(string s in deps) {
                        if(!depsList.Contains(s)) {
                            Trace.WriteLine(Trace.kGui, "Adding " + s);
                            depsList.Add(s);
                        }
                    }
                    n = n.NextPlugin;
                }
            }
        }
        
        foreach(string s in depsList) {
            try {
                Package p = PackageManager.Instance.PackageByNamespace(s);
                if(p.Namespace != Namespace) {
                    Assert.Check(p.Filename != "");
                    iPackageList.Add(Path.GetFileName(p.Filename));
                }
            } catch(PackageNotFound) {
                System.Console.WriteLine("WARNING: Could not find package for namespace {0}", s);
            }
        }
    }
    
    private const uint kVersion = 8;
    private string iFilename = "";
    private string iNamespace = "None";
    private string iType = "Unknown";
    private List<Plugin> iPluginList = new List<Plugin>();
    private List<string> iPackageList = new List<string>();
    private NodeHit iRoot = null;
}

} // Resources
} // Gui
} // Linn

File: Public/Gui/Linn/Gui/Resources/Package.py
import os
import Linn.Gui.Serialize
import Linn.Gui.XmlStream
import Linn.Gui.Factory
import Linn.Gui.Resources.Plugin
import Linn.Gui.SceneGraph.Visitor
import PackageManager
import Configuration

class InvalidVersion(Exception):
    def __init__(self, aVersion):
        self.iVersion = aVersion
    def __str__(self):
        return 'Invalid file version: Found %d expected %d' % (self.iVersion, Package.iVersion)

class DuplicatePluginName(Exception):
    def __init__(self, aName):
        self.iName = aName
    def __str__(self):
        return 'Error: Duplicate plugin name %s found' % self.iName
    

class Package(Linn.Gui.Serialize.ISerialize):
    iVersion = 6
    def __init__(self, aFilename=None, aType=None, aNamespace=None):
        if aFilename:
            self.iFilename = aFilename
        else:
            self.iFilename = None
        self.iNamespace = aNamespace
        self.iType = aType
        self.iPluginList = []  # holds references to all plugins currently in package
        self.iPackageList = [] # holds names of all packages this package references
        self.iRoot = None
        
        if self.iFilename:
            xmlFile = Linn.Gui.XmlStream.XmlStream()
            try:
                xmlFile.Open(Configuration.ProjectPath() + self.iFilename, Linn.Gui.XmlStream.kRead)
            except:
                print 'Trying default path...'
                xmlFile.Open(Configuration.DefaultPath() + self.iFilename, Linn.Gui.XmlStream.kRead)
            self.Load(xmlFile)
            xmlFile.Close()
        
    def Filename(self):
        return self.iFilename
    
    def Namespace(self):
        return self.iNamespace
    
    def SetNamespace(self, aNamespace):
        self.iNamespace = aNamespace
        for plugin in self.iPluginList:
            plugin.SetNamespace(aNamespace)
    
    def Type(self):
        return self.iType
    
    def AddPlugin(self, aPlugin, aInitialPlugin=None):
        aPlugin.SetNamespace(self.iNamespace)
        self.iPluginList.append(aPlugin)
        plugin = aPlugin.Plugin()
        if plugin and aInitialPlugin != plugin:
            self.AddPlugin(plugin, aInitialPlugin)
            
    def DeletePlugin(self, aPlugin, aInitialPlugin=None):
        pluginList = [aPlugin]
        if aPlugin.Namespace() == self.iNamespace:
            self.iPluginList.remove(aPlugin)
        plugin = aPlugin.Plugin()
        if plugin and aInitialPlugin != plugin:
            pluginList.extend(self.DeletePlugin(plugin, aInitialPlugin))
        return pluginList
            
    def AddNode(self, aNode):
        self.AddPlugin(aNode, aNode)
        if self.iRoot == None:
            #print 'Setting', aNode.Name(), 'as root'
            self.iRoot = aNode
        else:
            # ensure that the newly added polygon is on top of all current polygons
            zSearch = Linn.Gui.SceneGraph.Visitor.FindGreatestZ()
            z = zSearch.Search(self.iRoot)
            v = aNode.LocalSrt().Translation()
            v.iZ = z + 1
            aNode.SetLocalTranslation(v)
            self.iRoot.AddChild(aNode)
            
    def RemoveNode(self, aNode):
        pluginList = []
        aNode.Parent().RemoveChild(aNode)
        pluginList.extend(self.DeletePlugin(aNode, aNode))
        numChildren = aNode.NumChildren()
        for i in range(numChildren):
            pluginList.extend(self.RemoveNode(aNode.Child(numChildren - 1 - i)))
        return pluginList
            
    def Root(self):
        return self.iRoot
            
    def SetRoot(self, aPlugin):
        self.iRoot = aPlugin
    
    def PluginList(self):
        return self.iPluginList
    
    def PluginByName(self, aPlugin):
        # extract name and namespace from plugin's fullname
        index = aPlugin.find('.')
        if index:
            ns = aPlugin[:index]
            name = aPlugin[index+1:]
            if str(self.iNamespace) == ns:
                for plugin in self.iPluginList:
                    if plugin.Name() == name:
                        return plugin
        raise Linn.Gui.Resources.Plugin.PluginNotFound(aPlugin)

    def PackageList(self):
        return self.iPackageList
    
    def CalculatePackageReferences(self):
        """From the package's plugins obtain a list of external
            packages that this package is dependent on."""
        namespaceList = []
        for plugin in self.iPluginList:
            deps = plugin.PackageDependencies()
            for i in deps:
                try:
                    namespaceList.index(i)
                except Exception, e:
                    print 'Adding', i
                    namespaceList.append(i)
            nextPlugin = plugin.Plugin()
            while nextPlugin and nextPlugin != plugin:
                if nextPlugin.Namespace() != self.Namespace():
                    deps = nextPlugin.PackageDependencies()
                    for i in deps:
                        try:
                            namespaceList.index(i)
                        except Exception, e:
                            print 'Adding', i
                            namespaceList.append(i)
                nextPlugin = nextPlugin.Plugin()
                
        self.iPackageList = []
        for package in PackageManager.GetPackageManager().PackageList():
            try:
                index = namespaceList.index(package.Namespace())
                namespaceList.remove(namespaceList[index])
                if package.Namespace() != self.Namespace():
                    self.iPackageList.append(package.Filename())
            except Exception, e:
                pass
            
        for namespace in namespaceList:
            print 'WARNING: Could not find package for namespace %s' % namespace
    
    def Load(self, aStream):
        # obtain a list of all referenced packages
        package = aStream.LoadNode('Package')
        assert(len(package)==1)
        aStream.PushContextNode(package[0])
        self.iType = package[0].attributes[(None,'type')].value
        
        #print 'Loading package', self.iFilename
        #print '    Type', self.iType
        
        version = aStream.LoadNode('Version')
        if len(version) != 1:
            raise Linn.Gui.Serialize.InvalidLayoutFile('<Version>', len(version), 1)
        self.iVersion = int(version[0].firstChild.nodeValue)
        if self.iVersion != Package.iVersion:
            raise InvalidVersion(self.iVersion)
        
        namespace = aStream.LoadNode('Namespace')
        if len(namespace) != 1:
            raise Linn.Gui.Serialize.InvalidLayoutFile('<Namespace>', len(namespace), 1)
        self.iNamespace = namespace[0].firstChild.nodeValue
        
        packages = aStream.LoadNode('Package')
        #print '    %d package references' % len(packages)
        for package in packages:
            self.iPackageList.append(package.firstChild.nodeValue)
            
        # load in all the plugins
        plugins = aStream.LoadNode('Plugin')
        #print '    %d plugins' % len(plugins)
        for plugin in plugins:
            classType = plugin.attributes[(None,'class')].value
            aStream.PushContextNode(plugin)
            newPlugin = Linn.Gui.Factory.CreatePlugin(self.iNamespace, classType, aStream)
            aStream.PopContextNode()
            self.iPluginList.append(newPlugin)
            if classType == 'NodeHit' and newPlugin.Parent() == None:
               self.SetRoot(newPlugin) 
        aStream.PopContextNode()
    
    def Save(self, aStream):
        #print 'Saving package', aStream.Filename()
        self.iFilename = aStream.Filename()
        self.iFilename = self.iFilename.replace(Configuration.ProjectPath(), '')
        #print '    Type', self.iType
        package = aStream.CreateElement('Package')
        type = aStream.CreateAttribute('type', self.iType)
        package.setAttributeNodeNS(type)
        aStream.PushContextNode(package)
        
        version = aStream.CreateElement('Version')
        version.appendChild(aStream.CreateTextNode(Package.iVersion))
        aStream.SaveNode(version)
        
        namespace = aStream.CreateElement('Namespace')
        namespace.appendChild(aStream.CreateTextNode(self.iNamespace))
        aStream.SaveNode(namespace)
        
        self.CalculatePackageReferences()
        
        for packageFilename in self.iPackageList:
            packageRef = aStream.CreateElement('Package')
            packageRef.appendChild(aStream.CreateTextNode(packageFilename))
            aStream.SaveNode(packageRef)
        #print '    %d package references' % len(self.iPackageList)
        
        for plugin in self.iPluginList:
            pluginNode = aStream.CreateElement('Plugin')
            classType = aStream.CreateAttribute('class', None)
            pluginNode.setAttributeNodeNS(classType)
            aStream.PushContextNode(pluginNode)
            plugin.Save(aStream)
            aStream.PopContextNode()
            aStream.SaveNode(pluginNode)
        #print '    %d plugins' % len(self.iPluginList)
            
        aStream.PopContextNode()
        aStream.SaveNode(package)
        
    def Link(self):
        # check for duplicate plugin names
        numPlugins = len(self.iPluginList)
        for i in range(numPlugins):
            plugin = self.iPluginList[i]
            for j in range(i+1, numPlugins):
                testPlugin = self.iPluginList[j]
                if plugin.Fullname() == testPlugin.Fullname():
                    raise DuplicatePluginName(plugin.Fullname())
                
        for plugin in self.iPluginList:
            plugin.Link()
    
    def Refresh(self):
        pass
    
    
    